### PR TITLE
Bugfix/improvement for triangle-triangle intersect

### LINKED ISSUE #261: Possible bug in triangle-triangle intersection algorithm
A user-provided mesh was flagging some triangles as self-intersecting.

Based on visual inspection, the triangles do not appear to intersect.

image

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- Fixes #261 
- It exposes the tolerance threshold `EPS` for determining triangle intersections to the user.
- It sets the default tolerance for triangle intersection to `1E-8`. Previously, the intersections were getting a default of `1E-12` in the fuzzy comparison tests.
- This tolerance option is also exposed to users in the `mesh_tester` app
- I've also added a regression test derived from the mesh in #261

### CODE REVIEW

File: src/axom/primal/operators/detail/intersect_impl.hpp [hunk 1]
```diff
@@ -381,36 +386,20 @@ inline bool intersectTwoPermutedTriangles(const Point3& p1,
...
-  {
-    return false;
-  }
-
-  return true;
+  const bool bdr = includeBoundary;
+  /* *INDENT-OFF* */
+  return isLpeq(Vector3(q1, q2).dot(Triangle3(q1, p2, p1).normal()), 0.0, bdr, EPS)
+      && isLpeq(Vector3(p1, r2).dot(Triangle3(p1, p2, r1).normal()), 0.0, bdr, EPS);
+  /* *INDENT-ON* */
```
Comment (kennyweiss) [inline src/axom/primal/operators/detail/intersect_impl.hpp:400]: Here is the key change -- we need to get the user's `EPS` value down to this function.

File: src/tools/mesh_tester.cpp [hunk 2]
```diff
@@ -166,9 +162,17 @@ void Input::parse(int argc, char** argv, CLI::App& app)
                  "Distance threshold for welding vertices.")
   ->capture_default_str();
 
+  app.add_option("--intersectionThresh", intersectionThreshold,
+                 "Tolerance threshold to use when testing for intersecting triangles")
+  ->capture_default_str();
```
Comment (kennyweiss) [inline src/tools/mesh_tester.cpp:167]: Other key change -- the tolerance value is exposed to `mesh_tester` users.

**Comment by kennyweiss** [comment]
Update: In addition to azuer, this PR passed all internal `toss3` and `blueos` tests.

**Comment by mclarsen** [comment]
I have learned this lesson several times. I have a code where the entire bounds of the mesh is `1e-5`.  My suggestion is to create a relative epsilon based on extents of the coordinate system `eps = bounds.magnitude() * 1e-8`. Leaving the epsilonn up to the user will probably end up with the same issue because they might not be aware of internal usage of the epsilon. Alternatively, create a function that gives them a epsilon based on the mesh bounds.

**Comment by gzagaris** [comment]
I think we can have the `mesh_tester` calculate an `epsilon` in a pre-processing step, as @mclarsen indicated and use that if the user has not provided an `epsilon` -- That should help.

**Comment by kennyweiss** [comment]
Thanks @mclarsen -- I was thinking about adding something along these lines -- either a value that is relative to the bounding box (as you suggested), or perhaps to a percentage of the length of the shortest nondegenerate edge in the mesh.

I'll take care of it in a separate PR since it might be a bit tricky to get this right, and the bugfix here will at least get this user past their current problem.

**Pull request merged**